ui/signals.py

```diff
@@ -46,18 +48,37 @@ def delete_youtube_video(sender, **kwargs):
         remove_youtube_video.delay(kwargs['instance'].id)
 
 
-@receiver(post_save, sender=Video)
-def sync_youtube(sender, **kwargs):
+@receiver(post_save, sender=Collection)
+def update_collection_youtube(sender, **kwargs):
     """
+    If a collection's stream source is changed, sync YoutubeVideo objects for public Videos
+    """
+    if settings.ENABLE_VIDEO_PERMISSIONS:
+        for video in kwargs['instance'].videos.filter(is_public=True):
+            sync_youtube(video)
 
-    Upload a video to youtube if it is public and not already on YouTube.
-    Delete from youtube if it is there and permissions are not public.
+
+@receiver(post_save, sender=Video)
+def update_video_youtube(sender, **kwargs):
+    """
+    If a video's is_public field is changed, sync associated YoutubeVideo object
     """
     if settings.ENABLE_VIDEO_PERMISSIONS:
-        video = kwargs['instance']
-        yt_video_id = video.youtube_id
-        if video.is_public:
-            if not yt_video_id:
+        sync_youtube(kwargs['instance'])
+
+
+def sync_youtube(video):
+    """
+    Upload a video to youtube if it is public, not already on YouTube, and collection stream_source != cloudfront.
+    Delete from youtube if it exists and permissions are not public or collection stream_source == cloudfront.
+    """
+    yt_video = video.youtubevideo if hasattr(video, 'youtubevideo') else None
+    if video.is_public and video.collection.stream_source != StreamSource.CLOUDFRONT:
+        if video.original_video is not None:
+            if yt_video is None:
+                upload_youtube_video.delay(video.id)
+            elif yt_video.status in (YouTubeStatus.FAILED, YouTubeStatus.REJECTED):
+                YouTubeVideo.objects.get(id=yt_video.id).delete()
                 upload_youtube_video.delay(video.id)
-        elif yt_video_id:
-            YouTubeVideo.objects.get(id=yt_video_id).delete()
+    elif yt_video is not None:
+        YouTubeVideo.objects.get(id=yt_video.id).delete()
```

Comment:
After the `delete()` above it calls `upload_youtube_video`, should we do the same here?

Reply:
It should only be deleted here, because if it gets to this line it means either the video is no longer public or the collection `stream_source` is set to `Cloudfront`, so it should be removed from YouTube.